projects/js-packages/connection/state/resolvers.jsx

```diff
@@ -34,8 +38,25 @@ const connectionResolvers = {
 			return hasAuthorization;
 		},
 		*fulfill( redirectUri ) {
-			const response = yield actions.fetchAuthorizationUrl( redirectUri );
-			yield actions.setAuthorizationUrl( response.authorizeUrl );
+			try {
+				const response = yield actions.fetchAuthorizationUrl( redirectUri );
+				yield actions.setAuthorizationUrl( response.authorizeUrl );
+			} catch ( e ) {
+				let errorMessage = null;
+				if ( e.hasOwnProperty( 'response' ) ) {
+					errorMessage =
+						e.response.hasOwnProperty( 'message' ) && e.response.message
+							? e.response.message
+							: e.response.code;
+				} else {
+					errorMessage = e.toString();
+				}
+
+				yield actions.setUserIsConnecting( false );
+				yield actions.setSiteIsRegistering( false );
+				yield actions.setRegistrationError( errorMessage );
+				throw e;
+			}
```

Comment:
I don't think this is the place to do all this things, mainly based on the separation of concerns that I did at the refactor.

This resolver is responsible only to handle/provide the `authorizationUrl`, if someone uses it for that purpose it will have side effects this way, based on that I suggest to reduce this part only to get the `authorizationUrl` and if some error happens, we set the `authorizationUrl` for empty value:

```javascript
try {
  const response = yield actions.fetchAuthorizationUrl( redirectUri );
  yield actions.setAuthorizationUrl( response.authorizeUrl );
} catch ( e ) {
  yield actions.setAuthorizationUrl( '' );
}
```

And we make sure that into `then` method  at `CONNECT_USER` control (https://github.com/Automattic/jetpack/blob/master/projects/js-packages/connection/state/controls.jsx#L30-L40) we handle the situation that `authorizationUrl` doesn't exists, something like this:

```javascript
if ( authorizationUrl ) {
  const redirect = redirectFunc || ( url => window.location.assign( url ) );

  const url = new URL( authorizationUrl );

  if ( from ) {
    url.searchParams.set( 'from', encodeURIComponent( from ) );
  }

  const finalUrl = url.toString();
  redirect( finalUrl );
  resolve( finalUrl );
} else {
  reject( new Error( 'No authorizationUrl' ) );
}
```

And now we can handle this error into our `connectUser` action (https://github.com/Automattic/jetpack/blob/master/projects/js-packages/connection/state/actions.jsx#L57-L60):

```javascript
yield setUserIsConnecting( true );

try {
  yield { type: CONNECT_USER, from, redirectFunc };
} catch ( error ) {
  yield setUserIsConnecting( false );
  yield setRegistrationError( error );
}
```

This way we could avoid side-effects, and re-use things without dependency between them.